### ISSUE TITLE
Implement Intelligent Micro-Lesson Management System with Automated Promotion

### ISSUE THREAD

**Issue by Shredvardson** [Shredvardson/dl-starter#267]
## Problem Statement

Currently, micro-lessons serve as passive knowledge stores in `docs/micro-lessons/` with no automated mechanism to:

- Detect duplicate or semantically similar learnings across the growing knowledge base
- Identify high-value patterns that should be promoted to active enforcement mechanisms
- Track learning reuse and application in actual development workflows
- Suggest consolidation opportunities for related insights
- Promote frequently-used learnings to CI enforcement, documentation, Skills, or proactive reminders

This results in:
- Knowledge fragmentation and duplication
- Manual overhead in identifying promotion candidates
- Missed opportunities to automate preventable issues
- No feedback loop between learning capture and enforcement
- Lack of metrics on learning effectiveness and reuse

## Proposed Solution

Build an intelligent micro-lesson management system that automates the promotion lifecycle from passive knowledge capture to active enforcement. The system leverages:

1. **Semantic duplicate detection** using lightweight NLP (sentence embeddings)
2. **Usage pattern analysis** based on `UsedBy` counters and `Severity` metadata
3. **Automated promotion recommendations** with clear justification
4. **Feedback loop integration** tracking learning application in commits/PRs
5. **Continuous improvement dashboard** via `pnpm learn:analyze` command

#### 1. Semantic Duplicate Detection

- Analyze micro-lessons using semantic similarity (sentence embeddings)
- Flag duplicates when similarity threshold >80%
- Suggest consolidation with merged insights
- Use lightweight approach (no heavy ML frameworks like TensorFlow)
- Implementation options:
  - `sentence-transformers` via Python bridge (lightweight model)
  - `natural` NPM package for basic NLP
  - OpenAI embeddings API (token-efficient)

#### 2. Automated Promotion Analysis

Promotion paths based on usage patterns:

**To CI Enforcement:**
- Criteria: `UsedBy >= 3` AND automatable pattern
- Action: Generate validation script in `.lefthook/` or CI workflow
- Example: Database naming conventions → lint rule

**To Canonical Documentation:**
- Criteria: `Severity: high` AND `UsedBy >= 2`
- Action: Update relevant docs section with permanent guidance
- Example: Security best practices → docs/security.md

**To Skill:**
- Criteria: `UsedBy >= 5` AND multi-step workflow
- Action: Create structured Skill with token-efficient context
- Example: Complex debugging workflow → Skill template

**To Proactive Reminder:**
- Criteria: `UsedBy >= 3` AND context-specific
- Action: Add reminder to relevant command outputs
- Example: Migration safety checks → `db:migration:create` output

#### 3. Feedback Loop Integration

**Git Hook Integration:**
```typescript
// .lefthook/commit-msg/track-learning-usage.ts
// Scan commit messages/diffs for micro-lesson references
// Auto-increment UsedBy counter when learning is applied
```

**CI Integration:**
```yaml
# .github/workflows/learning-analysis.yml
# Weekly run to analyze promotion opportunities
# Flag learnings that have reached promotion thresholds
```

**Metrics Tracking:**
- Learning application rate (how often learnings are referenced)
- Time-to-enforcement (capture → automated enforcement)
- Reuse rate (percentage of learnings used ≥2 times)
- Consolidation rate (duplicate detection accuracy)

#### 4. Continuous Improvement Dashboard

**Command: `pnpm learn:analyze`**

Output format:
```
📊 Micro-Lesson Analysis Report
============================

🔄 PROMOTION RECOMMENDATIONS (3)

1. ML-008: Always validate RLS policies before deployment
   Current: UsedBy=4, Severity=high
   Recommendation: Promote to CI enforcement
   Justification: High severity + frequent use → automate validation
   Suggested Action: Create .lefthook/pre-push/validate-rls.ts

2. ML-015: Use absolute paths in agent prompts
   Current: UsedBy=6, Severity=medium
   Recommendation: Promote to Skill
   Justification: Multi-step workflow used frequently
   Suggested Action: Create skills/agent-prompting/absolute-paths.md

⚠️  DUPLICATE DETECTION (2)

1. ML-003 & ML-012 are 87% similar
   Suggestion: Consolidate into single learning
   Common theme: Database migration safety

📊 METRICS
- Total learnings: 24
- Reuse rate: 42% (10/24 used ≥2 times)
- Avg time-to-enforcement: 12 days
- Promotion candidates: 5
```

#### 5. Smart Indexing Enhancement

Enhance `INDEX.md` generation with promotion status:

```markdown
## Micro-Lessons Index

| ID | Title | Status | Used | Severity |
|----|-------|--------|------|----------|
| ML-001 | Always validate... | ✅ Enforced | 8 | high |
| ML-002 | Use semantic... | 📚 In Docs | 5 | medium |
| ML-003 | Implement retry... | ⚡ Reminder | 4 | medium |
| ML-004 | Check RLS before... | 🔄 Pending | 3 | high |
```

Status indicators:
- 🔄 Pending promotion (meets criteria, not yet promoted)
- ✅ Enforced in CI (automated validation)
- 📚 In canonical docs (permanent documentation)
- ⚡ Proactive reminder (added to command outputs)
- 💤 Dormant (UsedBy < 2, may need review)

### Core Functionality
- [ ] Semantic similarity detection identifies duplicate/similar learnings with >80% threshold
- [ ] `pnpm learn:analyze` command generates actionable promotion recommendations
- [ ] Automated `UsedBy` tracking via git hook or CI integration
- [ ] Promotion logic correctly identifies candidates based on criteria matrix
- [ ] Enhanced INDEX.md shows promotion status indicators for each learning

### Integration
- [ ] Git hook tracks learning application in commit messages/diffs
- [ ] CI workflow runs weekly analysis and flags promotion opportunities
- [ ] Integration with existing `/ops:learning-capture` command preserved
- [ ] Compatible with current micro-lesson template format

### Documentation
- [ ] Promotion criteria matrix documented (when to promote where)
- [ ] Workflow guide: from learning capture → analysis → promotion → enforcement
- [ ] Configuration documentation for similarity thresholds and promotion rules
- [ ] Examples of each promotion path (CI, docs, Skill, reminder)

### Success Metrics (2 weeks post-deployment)
- [ ] Detect at least 2 duplicate learnings and suggest consolidation
- [ ] Promote at least 1 high-value learning to CI enforcement
- [ ] Reduce time from learning capture to enforcement by 50%
- [ ] Achieve ≥30% learning reuse rate (learnings used ≥2 times)
- [ ] Zero false positives in duplicate detection

### Implementation Approach

**Phase 1: Semantic Analysis (6-8 hours)**
- Research and select lightweight NLP approach
- Implement similarity detection algorithm
- Create consolidation suggestion logic
- Add tests for edge cases (very similar vs identical)

**Phase 2: Promotion Logic (4-6 hours)**
- Implement criteria-based promotion recommendation engine
- Create justification text generation
- Build promotion action templates (CI scripts, doc updates, etc.)
- Add configuration for customizable thresholds

**Phase 3: Feedback Loop (3-4 hours)**
- Create git hook for tracking learning usage
- Implement `UsedBy` auto-increment logic
- Add metrics collection and storage
- Create CI workflow for weekly analysis

**Phase 4: Dashboard & Reporting (3-4 hours)**
- Build `learn:analyze` command
- Create formatted output with recommendations
- Enhance INDEX.md generation with status indicators
- Add interactive mode for immediate action on recommendations

### Technical Constraints

- **Lightweight implementation:** No TensorFlow, PyTorch, or heavy ML frameworks
- **Node/TypeScript ecosystem:** Use existing infrastructure (TypeScript, Node.js)
- **Template compatibility:** Work with current `docs/micro-lessons/TEMPLATE.md` format
- **Token efficiency:** Progressive disclosure for Skills promotion (avoid bloat)
- **Performance:** Similarity analysis should complete in <5 seconds for 100 learnings
- **Dependency management:** Minimize new dependencies, prefer built-in or lightweight packages

### Integration Points

- **Existing systems:**
  - `/ops:learning-capture` command (preserve workflow)
  - `docs/micro-lessons/` structure (read metadata)
  - Skills promotion workflow (output format)
  - CI validation scripts (template generation)

- **Git hooks:** `.lefthook/commit-msg/` for usage tracking
- **CI workflows:** `.github/workflows/` for automated analysis
- **Command system:** `commands/` directory for new `learn:analyze` command

### NLP Options Evaluation

**Option 1: sentence-transformers (Python bridge)**
- Pros: High accuracy, established library
- Cons: Python dependency, slower startup
- Model: `all-MiniLM-L6-v2` (small, fast)

**Option 2: OpenAI Embeddings API**
- Pros: No local dependencies, excellent accuracy
- Cons: API costs, rate limits
- Recommended for production use

**Option 3: natural NPM package**
- Pros: Pure JavaScript, no external deps
- Cons: Lower accuracy, manual tuning needed
- Good for MVP/testing

**Recommendation:** Start with OpenAI embeddings (already used in project), fallback to `natural` for offline use.

### Data Storage

- Similarity cache: `docs/micro-lessons/.similarity-cache.json`
- Metrics history: `docs/micro-lessons/.metrics.json`
- Promotion log: `docs/micro-lessons/.promotion-log.json`
- Git-ignored for privacy, regenerated on demand

## Related Issues

- Research: 2025 KM trends (AI automation, semantic search, Kaizen loops)
- Related to: `/ops:learning-capture` command implementation
- Related to: Skills promotion workflow
- Depends on: Existing micro-lesson template and INDEX.md generation

## Future Enhancements

- Real-time similarity checking during learning capture
- Machine learning to improve promotion recommendations over time
- Integration with code review bots to suggest relevant learnings
- Automatic PR generation for promoted learnings
- Learning effectiveness scoring based on incident prevention
